### PR TITLE
Deprecate config item `max-memory` and add items `server-memory-quota` and `memory-usage-alarm-ratio`

### PR DESCRIPTION
**Pull request by wshwsh12**
### What is changed, added or deleted? (Required)

Deprecate config item `max-memory` and add items `server-memory-quota` and `memory-usage-alarm-ratio`

`max-memory` has been deprecated and use `server-memory-quota` instead now.
`memory-usage-alarm-ratio` is introduced in this pr https://github.com/pingcap/tidb/pull/18858.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: 
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Have version specific changes 
- [ ] Might cause conflicts

### CODE REVIEW

File: tidb-configuration-file.md [hunk 1]
```diff
@@ -283,11 +283,18 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
...
 + 默认值为 0 表示使用机器上所有的 CPU；如果设置成 n，那么 TiDB 会使用 n 个 CPU 数量。
 
-### `max-memory`
+### `server-memory-quota`
 
-+ Prepare cache LRU 使用的最大内存限制。当 Prepare cache LRU 的内存使用超过 `performance.max-memory * (1 - prepared-plan-cache.memory-guard-ratio)` 时，会剔除 LRU 中的元素。
-+ 默认值：0
-+ 这个配置只有在 `prepared-plan-cache.enabled` 为 `true` 的情况才会生效。当 LRU 的 size 大于 `prepared-plan-cache.capacity` 时，也会剔除 LRU 中的元素。
++ TiDB 内存使用限制。
++ 默认值：0 （Byte）
```
Comment (yikeke) [inline tidb-configuration-file.md:289]:
```suggestion
+ TiDB 内存使用限制，单位为字节。
+ 默认值：0 
```

**Review by yikeke — APPROVED** [review]
LGTM

**Comment by yikeke** [comment]
@wshwsh12 Please help request a tech review.

File: tidb-configuration-file.md [hunk 2]
```diff
@@ -283,11 +283,18 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
...
-+ 这个配置只有在 `prepared-plan-cache.enabled` 为 `true` 的情况才会生效。当 LRU 的 size 大于 `prepared-plan-cache.capacity` 时，也会剔除 LRU 中的元素。
++ TiDB 内存使用限制，单位为字节。
++ 默认值：0 
++ 默认值为 0 表示无内存限制。
+
+### `memory-usage-alarm-ratio`
+
++ TiDB 内存使用报警阈值。
++ 默认值：0.8
++ 该配置项的有效范围为 0 到 1。 如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，则会将相关信息记录到目录 `tmp-storage-path/record` 中。
```
Comment (XuHuaiyu) [inline tidb-configuration-file.md:296]: 1. 这个太简单了，得写下会记录哪些信息。
2. 此外，关于tmp-storage-path 得加个链接到对应的文档项上去
3. server-memory-quota 也需要加个链接到对应的文档项上
4. system memory size 换成中文

**Comment by yikeke** [comment]
Please resolve the conflicts. @wshwsh12

File: tidb-configuration-file.md [hunk 3]
```diff
@@ -283,11 +283,18 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
...
++ TiDB 内存使用限制，单位为字节。
++ 默认值：0 
++ 默认值为 0 表示无内存限制。
+
+### `memory-usage-alarm-ratio`
+
++ TiDB 内存使用报警阈值。
++ 默认值：0.8
++ 该配置项的有效范围为 0 到 1。 如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，我们认为 TiDB 目前存在 OOM 的风险，会将当前正在执行的所有 SQLs 中内存使用最高的 10 条 SQL 和运行时间最长的 10 条 SQL 以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中,并输出一条包含关键子 `the Tidb instance has the risk of OOM` 的日志。
++ 注意：如果配置项 [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) 被设置且大于 0，则内存报警阈值将为 `memory-usage-alarm-ratio * server-memory-quota`；否则，内存报警阈值将为 `memory-usage-alarm-ratio * 系统内存大小`。
```
Comment (CharLotteiu) [inline tidb-configuration-file.md:297]: 此处提到内存报警阈值为 `memory-usage-alarm-ratio` 与 `server-memory-quota` 两个配置项的值的乘积，但第 294 行对 `memory-usage-alarm-ratio` 这个配置项的解释也是“TiDB 内存使用报警阈值”，前后有点矛盾。而从配置项名称中的“ratio”来看，应该这个配置项的值规定的是内存使用达到何种比例时报警，是否考虑修改一下第 294 行对配置项的说明？

File: tidb-configuration-file.md [hunk 4]
```diff
@@ -283,11 +283,18 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
...
-+ 这个配置在 `prepared-plan-cache.enabled` 为 `true`（默认值）的情况才会生效。当 LRU 的 size 大于 `prepared-plan-cache.capacity` 时，也会剔除 LRU 中的元素。
++ TiDB 内存使用限制，单位为字节。
++ 默认值：0 
++ 默认值为 0 表示无内存限制。
+
+### `memory-usage-alarm-ratio`
+
++ TiDB 内存使用报警阈值。
++ 默认值：0.8
++ 该配置项的有效范围为 0 到 1。 如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，我们认为 TiDB 目前存在 OOM 的风险，会将当前正在执行的所有 SQLs 中内存使用最高的 10 条 SQL 和运行时间最长的 10 条 SQL 以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中,并输出一条包含关键子 `the Tidb instance has the risk of OOM` 的日志。
```
Comment (yikeke) [inline tidb-configuration-file.md:296]:
```suggestion
+ 该配置项的有效范围为 0 到 1。如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，则会认为存在 OOM 的风险，会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中，并输出一条包含关键字 `the Tidb instance has the risk of OOM` 的日志。
```

**Comment by CharLotteiu** [comment]
LGTM. @yikeke PTAL.

File: tidb-configuration-file.md [hunk 5]
```diff
@@ -291,9 +291,9 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
 
 ### `memory-usage-alarm-ratio`
 
-+ TiDB 内存使用报警阈值。
++ TiDB 内存使用报警阈值占总内存的比例。
```
Comment (yikeke) [inline tidb-configuration-file.md:294]:
```suggestion
+ TiDB 内存使用占总内存的比例超过一定阈值时会报警。该配置项的有效范围为 0 到 1。如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。
```

File: tidb-configuration-file.md [hunk 6]
```diff
@@ -291,9 +291,9 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
...
 + 默认值：0.8
-+ 该配置项的有效范围为 0 到 1。 如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，我们认为 TiDB 目前存在 OOM 的风险，会将当前正在执行的所有 SQLs 中内存使用最高的 10 条 SQL 和运行时间最长的 10 条 SQL 以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中,并输出一条包含关键子 `the Tidb instance has the risk of OOM` 的日志。
++ 该配置项的有效范围为 0 到 1。如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，则会认为存在 OOM 的风险，会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中，并输出一条包含关键字 `tidb-server has the risk of OOM` 的日志。
 + 注意：如果配置项 [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) 被设置且大于 0，则内存报警阈值将为 `memory-usage-alarm-ratio * server-memory-quota`；否则，内存报警阈值将为 `memory-usage-alarm-ratio * 系统内存大小`。
```
Comment (yikeke) [inline tidb-configuration-file.md:297]:
```suggestion
```

File: tidb-configuration-file.md [hunk 7]
```diff
@@ -291,9 +291,9 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
 
 ### `memory-usage-alarm-ratio`
 
-+ TiDB 内存使用报警阈值。
++ TiDB 内存使用报警阈值占总内存的比例。
 + 默认值：0.8
-+ 该配置项的有效范围为 0 到 1。 如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，我们认为 TiDB 目前存在 OOM 的风险，会将当前正在执行的所有 SQLs 中内存使用最高的 10 条 SQL 和运行时间最长的 10 条 SQL 以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中,并输出一条包含关键子 `the Tidb instance has the risk of OOM` 的日志。
++ 该配置项的有效范围为 0 到 1。如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，则会认为存在 OOM 的风险，会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中，并输出一条包含关键字 `tidb-server has the risk of OOM` 的日志。
```
Comment (yikeke) [inline tidb-configuration-file.md:296]:
```suggestion
+ 当内存阈值报警功能开启时，如果配置项 [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) 未设置，则内存报警阈值为 `memory-usage-alarm-ratio * 系统内存大小`；如果 `server-memory-quota` 被设置且大于 0，则内存报警阈值为 `memory-usage-alarm-ratio * server-memory-quota`。
+ 当 TiDB 检测到内存使用超过了阈值，则会认为存在 OOM 的风险，会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中，并输出一条包含关键字 `tidb-server has the risk of OOM` 的日志。
```

**Review by yikeke — COMMENTED** [review]
重新组织了下这三段的逻辑，希望这样更清楚些：

File: tidb-configuration-file.md [hunk 8]
```diff
@@ -283,11 +283,18 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
...
++ TiDB 内存使用限制，单位为字节。
++ 默认值：0 
++ 默认值为 0 表示无内存限制。
+
+### `memory-usage-alarm-ratio`
+
++ TiDB 内存使用报警阈值占总内存的比例。
++ 默认值：0.8
++ 该配置项的有效范围为 0 到 1。如果配置该选项为 0 或 1，则表示关闭内存阈值报警功能。否则，当 TiDB 检测到内存使用超过了阈值，则会认为存在 OOM 的风险，会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中，并输出一条包含关键字 `tidb-server has the risk of OOM` 的日志。
++ 注意：如果配置项 [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) 被设置且大于 0，则内存报警阈值将为 `memory-usage-alarm-ratio * server-memory-quota`；否则，内存报警阈值将为 `memory-usage-alarm-ratio * 系统内存大小`。
```
Comment (yikeke) [inline tidb-configuration-file.md:297]:
```suggestion
```

**Comment by yikeke** [comment]
Note: After https://github.com/pingcap/tidb/pull/20473 is merged, we need a 4.0 cherry-pick of this PR.

**Review by yikeke — APPROVED** [review]
LGTM

**Comment by yikeke** [comment]
Why closed it? @wshwsh12

**Comment by wshwsh12** [comment]
It needs more information(eg. independent page in troubleshooting) to introduce all memory config. 
The bot comments too frequently, so I close this pr now.
I will reopen this pr after I prepare all the information.

**Pull request closed without merging**